File: OlderCodes/Ccodes/numerov.c
/* 
************************************************************************
*  numerov.c: Numerov method to find eigenvalues and eigenfunctions of *
*       a particle in a potential well. As written this program        *
*       only finds odd eigenfunctions correctly and fails for          *
*       even eigenfunctions and if there is no eigenvalue in the       *
*             specified range. The final function is not normalized.   *
*                      *
*  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; & CC BORDEIANU, Univ Bucharest, 2008
   Support by National Science Foundation                              

************************************************************************
*/
#include <stdio.h>
#include <math.h>

#define steps 1000         /* half width of potential well */
#define V -0.001         /* depth of potential well */
#define eps 1E-8         /* accuracy for eigenvalues */
#define Emin -0.001        /* set to find the third */
#define Emax -0.00085        /* eigenfunction */

double k2(int i, double E);      /* returns potential at x */
double diff(double E);         /* difference of derivatives */
void plot(double E);               /* data for final plot */ 

main()
{
   double E, min, max;
   int i=0;          /* counter for iterations */
   min=Emin;        
   max=Emax;
  
   do
   {
      i++;  
      E=(max+min)/2.0;         /* divide energy range */
      if (diff(max)*diff(E)>0) max=E;   /* the bisection algorithm */
      else min=E;
   }while(fabs(diff(E))>eps);
   
   printf("eigenvalue E=%.10f\n", E);
   printf("after %d iterations\n", i);
   plot(E);       
}


/*----------------------- end of main program ------------------------*/ 

/* function returns difference between left and right wavefunction */
double diff(double E)
{
   double one, two, three, plus, minus;
   int i;
   
   one=0.0;
   two=0.00001;
   for (i=1; i<=1500; i++)    /* left side first */
   {
      three=(2*two*(1-5./12.*k2(i,E))-
          (1.+1./12.*k2(i-1,E))*one)/(1.+1./12.*k2(i+1,E));
      one=two;
      two=three;
   }
   minus=two;       /* value at matching point */ 

   one=0.0;       /* reset starting conditions */
   two=0.00001;
   for (i=1; i<500; i++)    /* now the right side */
   {
      three=(2*two*(1.-5./12.*k2(i,E))-
          (1.+1./12.*k2(i+1,E))*one)/(1.+1./12.*k2(i-1,E));
      one=two;
      two=three;
   }
   plus=two;              /* value at matching point */
   return((minus-plus));
}
/*--------------------------------------------------------------------*/

/* function returns k-vector at depending on position i */
double k2(int i, double E)    
{
   if (i<500) return(E);    /* outside the well */
   if (i>=500) return (E-(V));    /* inside the well */
}
/*--------------------------------------------------------------------*/

/* write data for eigenfuntion into files left.dat, right.dat */
void plot(double E)     
{         
   double one, two, three;
   int i;
   
   FILE *right, *left;      /* save data in files */
   right=fopen("right.dat", "w");
   left=fopen("left.dat", "w");
   
   one=0.0;
   two=0.00001;
   for (i=1; i<=1500; i++)    /* left side first */
   {
      three=(2*two*(1-5./12.*k2(i,E))-
          (1+1./12.*k2(i-1,E))*one)/(1+1./12.*k2(i+1,E));
      fprintf(left, "%d\t%f\n", i-1000, three/2.8);
      
      one=two;        
      two=three;
   }
   
   one=0.0;       /* reset starting conditions */
   two=0.00001;
   for (i=1; i<500; i++)    /* now the right side */
   {
      three=(2*two*(1.-5./12.*k2(i,E))-
          (1+1./12.*k2(i+1,E))*one)/(1.+1./12.*k2(i-1,E));
      fprintf(right, "%d\t%f\n", 1000-i, three/2.8);
      
      one=two;
      two=three;
   }
   fclose(left);
   fclose(right);
   printf("data saved in left.dat and right.dat\n");
}
   

File: OlderCodes/JavaCodes/Numerov.java
/* From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
   Support by National Science Foundation                              
   */
// Numerov.java solves Schroedinger eq via Numerov method (rk also OK)

public class Numerov  {
  int steps = 1000;                                   // Half width of potential well
  static double V = -0.001, eps = 1E-5, Emin = -0.001, Emax = -0.00085;    
  public static void main(String[] argv) {
    double E, min, max; 
    int i = 0;                                              // Counter for iterations
    min = Emin;  max = Emax;                 
    do  { 
      i++ ;   
      E = (max + min)/2. ;                                     // Divide energy range
      System.out.println("eigenvalue E = "+E+", diff(E) = " +diff(E)); 
      System.out.println("after " + i + " iterations"); 
      if ( diff(max)*diff(E)>0 ) max = E;  else min = E;                 // Bisection
    } 
    while(Math.abs(diff(E))>eps); 
    System.out.println("eigenvalue E = " + E); 
    System.out.println("after " + i + " iterations");   
  }
  
  public static double diff(double E)  {
    double ud[] = new double[1501], u[] = new double[1501], k2l[] = new double[1501];
		double k2r[] = new double[1501], xl0, xr0, xl, xr, plus = 0., minus = 0., h; 
    int i, im, nl, nr, n = 1501; 
    u[0] = 0. ;   u[1] = 0.00001; 
    xl0 = -1000;   xr0 = 1000; 
    h  = (xr0-xl0)/(n-1); 
    for ( i = 0;  i  <  n; ++ i)  { 
      xl = xl0 + i*h; 
      xr = xr0-i*h; 
      k2l[i] = (E-v(xl)); 
      k2r[i] = (E-v(xr));     
    }
    im = 500;                                                   // The matching point
    nl = im + 1; 
    nr = n-im + 1; 
    for ( i=1; i < nl-1; i++ )  { u[i + 1] = (2*u[i]*(1-5.*h*h/12.*k2l[i])
				                  - (1.+ h*h/12.*k2l[i-1])*u[i-1])/(1.+ h*h/12.*k2l[i+1]); }
    minus = (u[i + 1]-u[i])/h*u[i];                         // Logarithmic derivative
    u[0] = 0. ; u[1] = 0.00001; 
    for ( i=1; i < nr-1; i++ )  { u[i+1] = (2*u[i]*(1-5.*h*h/12.*k2r[i])-
                 (1.+ h*h/12.*k2r[i-1])*u[i-1])/(1.+ h*h/12.*k2r[i+1]); }
    plus = (u[i + 1]-u[i])/h*u[i];                          // Logarithmic derivative
    System.out.println ("minus-plus/minus + plus = " + ((minus-plus)/(minus+plus)));
    return((minus-plus)/(minus + plus)); 
  }
  
  public static double v(double x)  {   
    double v;       
    if (Math.abs(x) <= 500)  v = -0.001;   else v = 0; 
    return v; 
  }
}
